File: dabbit.Base/Channel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;

namespace dabbit.Base
{
    public class Channel
    {
        protected Channel()
        {

        }

        public Channel(Server svr)
        {
            this.ServerOf = svr;
        }

        public string Name { get; set; }
        public virtual Topic Topic { get; set; }
        public virtual List<Mode> Modes { get; internal set; }
        public virtual List<User> Users { get; internal set; }
        public string Display { get; internal set; }

        public DateTime Created { get; internal set; }

        public Server ServerOf { get; internal set; }

        public Channel DeepClone()
        {
            Channel chan = (Channel)this.MemberwiseClone();
            chan.Topic = this.Topic.DeepClone();
            
            return chan;
        }

        public bool ChannelLoaded
        {
            get
            {
                return !String.IsNullOrEmpty(Name) && Modes != null && Users.Count() != 0 && !String.IsNullOrEmpty(Display);
            }
        }

        public override string ToString()
        {
            return this.Name + "2s";
        }
    }

    public class Topic
    {
        public string Display { get; set; }
        public string SetBy { get; set; }
        public DateTime DateSet { get; set; }

        public Topic DeepClone()
        {
            return (Topic)this.MemberwiseClone();
        }
    }

}

File: dabbit.Base/Channel.js
var System = require("../System")(__dirname + "/../System");
var Server = {}; //require("./Server");
var Topic = require("./Topic");

function Channel(svr) {
    // Create an object
    System.Object.call(this, this);

    if (!svr || svr.GetType() != "Server") {
        throw new System.ArgumentException("Invalid Server Parameter svr");
    }

    this.Name = String.Empty;
    this.Modes = [];
    this.Users = [];

    this.__defineGetter__("Display", function(){
        return this.Name;
    });
    var display = String.Empty;

    this.__defineGetter__("Created", function(){
        return created;
    });
    var created = new Date();

    this.__defineGetter__("ServerOf", function(){
        return server.Value;
    });

    this.__defineGetter__("ChannelLoaded", function(){
        return !String.IsNullOrEmpty(this.Name) && this.Modes.length != 0 && this.Users.length != 0 && !String.IsNullOrEmpty(this.Display);
    });

    this.__defineGetter__("Topic", function(){
        return !String.IsNullOrEmpty(this.Name) && this.Modes.length != 0 && this.Users.length != 0 && !String.IsNullOrEmpty(this.Display);
    });
    this.__defineSetter__("Topic", function(){
        return !String.IsNullOrEmpty(this.Name) && this.Modes.length != 0 && this.Users.length != 0 && !String.IsNullOrEmpty(this.Display);
    });
    var topic = new System.Javascript.CheckedProperty(new Topic(), System.Typeof(Topic));

    this.toString = function()
    {
        return this.Name;
    }

    var server = new System.Javascript.CheckedProperty(svr, System.Typeof(Server));

}
System.Javascript.Inherit(System.Object, Channel);

module.exports = Channel;
